typescript/structural/decorator/controllers/CoffeeController.ts

```diff
@@ -0,0 +1,41 @@
+import checkout from '../views/checkout'
+import { render } from '../helpers/rendering'
+import Order from '../models/Order'
+import InstantCoffeeWithMilk from '../models/InstantCoffeeWithMilk'
+import InstantCoffeeWithSugar from '../models/InstantCoffeeWithSugar'
+
+export default class CoffeeController {
+  index() {
+    render('Taking orders...')
+
+    // Great! We can take coffee orders!
+    const order: Order = new Order
+
+    order.add(new InstantCoffeeWithMilk)
+    order.add(new InstantCoffeeWithSugar)
+
+    // TODO: Implement the decorator pattern:
+    // * Identify the primary component that has optional layers
+    // * Identify the methods in common to both the primary component and the optional layers
+    // * Create the component interface and declare the common methods
+    // * Create the concrete component class and define the base behavior
+    // * Create the base decorator class. It should have a property for referencing the base or decorated object
+    // * Create concrete decorators by extending them from the base decorator
+    // * Compose!
+
+    // A revolutionary customer decides they want Milk and Sugar with their coffee... wuuuttt?
+    // ...Should we create a new concrete class InstantCoffeeWithMilkAndSugar? Ah ah ah
```

Comment:
This is great as a next step - are you expecting that people would have already refactored the code at this point? If so, the reference to the concrete classes would have been changed already. If you want to highlight the point though, perhaps move this before the TODO above. Same with the other one below.

Reply:
Yes, I'm expecting they would have created the necessary interfaces and abstraction. When trialing my own task, for me I left those concretions hanging around and just composed the next two "coffee orders".
I like your approach, I think it's best practice to also refactor the existing code (I would in any given real project!), so why not!

Reply:
Love it :smile: